File: ruby/lib/fizzy/http.rb
# frozen_string_literal: true

require "faraday"
require "json"
require "time"
require "uri"

module Fizzy
  # HTTP client layer with retry, backoff, and caching support.
  # This is an internal class used by Client; you typically don't use it directly.
  class Http
    # Default User-Agent header
    USER_AGENT = "fizzy-sdk-ruby/#{VERSION} (api:#{API_VERSION})".freeze

    # @param config [Config] configuration settings
    # @param token_provider [TokenProvider, nil] token provider (deprecated, use auth_strategy)
    # @param auth_strategy [AuthStrategy, nil] authentication strategy
    # @param hooks [Hooks] observability hooks
    def initialize(config:, token_provider: nil, auth_strategy: nil, hooks: nil)
      @config = config
      @auth_strategy = auth_strategy || BearerAuth.new(token_provider)
      @token_provider = token_provider || (@auth_strategy.is_a?(BearerAuth) ? @auth_strategy.token_provider : nil)
      @hooks = hooks || NoopHooks.new
      @faraday = build_faraday_client
    end

    # @return [String] the configured base URL
    def base_url
      @config.base_url
    end

    # Performs a GET request.
    # @param path [String] URL path
    # @param params [Hash] query parameters
    # @return [Response]
    def get(path, params: {})
      request(:get, path, params: params)
    end

    # Performs a GET request to an absolute URL.
    # Used for endpoints not on the base API.
    # @param url [String] absolute URL
    # @param params [Hash] query parameters
    # @return [Response]
    def get_absolute(url, params: {})
      request(:get, url, params: params)
    end

    # Performs a POST request.
    # @param path [String] URL path
    # @param body [Hash, nil] request body
    # @return [Response]
    def post(path, body: nil)
      request(:post, path, body: body)
    end

    # Performs a PUT request.
    # @param path [String] URL path
    # @param body [Hash, nil] request body
    # @return [Response]
    def put(path, body: nil)
      request(:put, path, body: body)
    end

    # Performs a PATCH request.
    # @param path [String] URL path
    # @param body [Hash, nil] request body
    # @return [Response]
    def patch(path, body: nil)
      request(:patch, path, body: body)
    end

    # Performs a DELETE request.
    # @param path [String] URL path
    # @return [Response]
    def delete(path)
      request(:delete, path)
    end

    # Performs a POST request with raw binary data.
    # Used for file uploads.
    # @param path [String] URL path
    # @param body [String, IO] raw binary data
    # @param content_type [String] MIME content type
    # @return [Response]
    def post_raw(path, body:, content_type:)
      url = build_url(path)
      single_request_raw(:post, url, body: body, content_type: content_type, attempt: 1)
    end

    # Fetches all pages of a paginated resource.
    # @param path [String] initial URL path
    # @param params [Hash] query parameters
    # @yield [Hash] each item from the response
    # @return [Enumerator] if no block given
    def paginate(path, params: {}, &block)
      return to_enum(:paginate, path, params: params) unless block

      base_url = build_url(path)
      url = base_url
      page = 0

      loop do
        page += 1
        break if page > @config.max_pages

        @hooks.on_paginate(url, page)
        response = get(url, params: page == 1 ? params : {})

        Security.check_body_size!(response.body, Security::MAX_RESPONSE_BODY_BYTES)

        begin
          items = JSON.parse(response.body)
        rescue JSON::ParserError => e
          raise Fizzy::APIError.new(
            "Failed to parse paginated response (page #{page}): #{Security.truncate(e.message)}"
          )
        end
        items.each(&block)

        next_url = parse_next_link(response.headers["Link"])
        break if next_url.nil?

        next_url = Security.resolve_url(url, next_url)

        unless Security.same_origin?(next_url, base_url)
          raise Fizzy::APIError.new(
            "Pagination Link header points to different origin: #{Security.truncate(next_url)}"
          )
        end

        url = next_url
      end
    end

    private

    def build_faraday_client
      Faraday.new(url: @config.base_url) do |f|
        f.options.timeout = @config.timeout
        f.options.open_timeout = 10
        f.request :json
        f.response :raise_error
        f.adapter Faraday.default_adapter
      end
    end

    def request(method, path, params: {}, body: nil)
      url = build_url(path)

      # Mutations don't retry on 429/5xx to avoid duplicating data
      if method == :get
        request_with_retry(method, url, params: params)
      else
        single_request(method, url, params: params, body: body, attempt: 1)
      end
    end

    def request_with_retry(method, url, params: {})
      attempt = 0
      last_error = nil

      loop do
        attempt += 1
        break if attempt > @config.max_retries

        begin
          return single_request(method, url, params: params, body: nil, attempt: attempt)
        rescue Fizzy::RateLimitError, Fizzy::NetworkError, Fizzy::APIError => e
          raise e unless e.retryable?

          last_error = e

          # Don't sleep if this was the last attempt
          break if attempt >= @config.max_retries

          delay = calculate_delay(attempt, e.retry_after)

          @hooks.on_retry(
            RequestInfo.new(method: method.to_s.upcase, url: url, attempt: attempt),
            attempt + 1, e, delay
          )
          sleep(delay)
        end
      end

      raise last_error || Fizzy::APIError.new("Request failed after #{@config.max_retries} retries")
    end

    def single_request(method, url, params:, body:, attempt:, retry_count: 0)
      info = RequestInfo.new(method: method.to_s.upcase, url: url, attempt: attempt)
      @hooks.on_request_start(info)

      start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

      begin
        response = @faraday.run_request(method, url, body, request_headers) do |req|
          req.params.merge!(params) if params.any?
        end

        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        result = RequestResult.new(status_code: response.status, duration: duration)
        @hooks.on_request_end(info, result)

        Response.new(
          body: response.body,
          status: response.status,
          headers: response.headers
        )
      rescue Faraday::ServerError, Faraday::ClientError => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = handle_error(e)
        result = RequestResult.new(
          status_code: e.response&.dig(:status),
          duration: duration,
          error: error,
          retry_after: error.respond_to?(:retry_after) ? error.retry_after : nil
        )
        @hooks.on_request_end(info, result)

        # After a successful token refresh on 401, retry the request once
        if error.is_a?(Fizzy::AuthError) && error.http_status == 401 && retry_count < 1 && @token_refreshed
          @token_refreshed = false
          return single_request(method, url, params: params, body: body, attempt: attempt, retry_count: retry_count + 1)
        end

        raise error
      rescue Faraday::Error => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = Fizzy::NetworkError.new("Connection failed", cause: e)
        result = RequestResult.new(duration: duration, error: error)
        @hooks.on_request_end(info, result)
        raise error
      end
    end

    def request_headers
      headers = {
        "User-Agent" => USER_AGENT,
        "Accept" => "application/json"
      }
      @auth_strategy.authenticate(headers)
      headers
    end

    def single_request_raw(method, url, body:, content_type:, attempt:)
      info = RequestInfo.new(method: method.to_s.upcase, url: url, attempt: attempt)
      @hooks.on_request_start(info)

      start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

      begin
        headers = request_headers.merge("Content-Type" => content_type)
        response = @faraday.run_request(method, url, body, headers)

        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        result = RequestResult.new(status_code: response.status, duration: duration)
        @hooks.on_request_end(info, result)

        Response.new(
          body: response.body,
          status: response.status,
          headers: response.headers
        )
      rescue Faraday::ServerError, Faraday::ClientError => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = handle_error(e)
        result = RequestResult.new(
          status_code: e.response&.dig(:status),
          duration: duration,
          error: error,
          retry_after: error.respond_to?(:retry_after) ? error.retry_after : nil
        )
        @hooks.on_request_end(info, result)
        raise error
      rescue Faraday::Error => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = Fizzy::NetworkError.new("Connection failed", cause: e)
        result = RequestResult.new(duration: duration, error: error)
        @hooks.on_request_end(info, result)
        raise error
      end
    end

    def handle_error(error)
      status = error.response&.dig(:status)
      body = error.response&.dig(:body)
      headers = error.response&.dig(:headers) || {}

      retry_after = parse_retry_after(headers["Retry-After"] || headers["retry-after"])
      request_id = headers["X-Request-Id"] || headers["x-request-id"]

      err = case status
      when 401
        # Try token refresh; flag for caller to retry
        @token_refreshed = @token_provider&.refreshable? && @token_provider.refresh
        Fizzy::AuthError.new("Authentication failed")
      when 403
        Fizzy::ForbiddenError.new("Access denied")
      when 404
        Fizzy::NotFoundError.new("Resource", "unknown")
      when 429
        Fizzy::RateLimitError.new(retry_after: retry_after)
      when 400, 422
        message = Security.truncate(Fizzy.parse_error_message(body) || "Validation failed")
        Fizzy::ValidationError.new(message, http_status: status)
      when 500
        Fizzy::APIError.new("Server error (500)", http_status: 500, retryable: true)
      when 502, 503, 504
        Fizzy::APIError.new("Gateway error (#{status})", http_status: status, retryable: true)
      else
        message = Security.truncate(Fizzy.parse_error_message(body) || "Request failed (HTTP #{status})")
        Fizzy::APIError.from_status(status || 0, message)
      end

      err.instance_variable_set(:@request_id, request_id) if request_id
      err
    end

    def build_url(path)
      if path.start_with?("https://")
        return path
      elsif path.start_with?("http://")
        raise Fizzy::UsageError.new("URL must use HTTPS: #{path}")
      end

      path = "/#{path}" unless path.start_with?("/")
      "#{@config.base_url}#{path}"
    end

    def calculate_delay(attempt, server_retry_after)
      return server_retry_after if server_retry_after&.positive?

      # Exponential backoff: base_delay * 2^(attempt-1) + jitter
      base = @config.base_delay * (2**(attempt - 1))
      jitter = rand * @config.max_jitter
      base + jitter
    end

    def parse_retry_after(value)
      return nil if value.nil? || value.empty?

      # Try parsing as seconds (integer)
      seconds = Integer(value, exception: false)
      return seconds if seconds&.positive?

      # Try parsing as HTTP-date
      begin
        date = Time.httpdate(value)
        diff = (date - Time.now).to_i
        return diff if diff.positive?
      rescue ArgumentError
        # Not a valid HTTP-date
      end

      nil
    end

    def parse_next_link(link_header)
      return nil if link_header.nil? || link_header.empty?

      link_header.split(",").each do |part|
        part = part.strip
        next unless part.include?('rel="next"')

        match = part.match(/<([^>]+)>/)
        return match[1] if match
      end

      nil
    end
  end

  # Wraps an HTTP response.
  class Response
    # @return [String] response body
    attr_reader :body

    # @return [Integer] HTTP status code
    attr_reader :status

    # @return [Hash] response headers
    attr_reader :headers

    def initialize(body:, status:, headers:)
      @body = body
      @status = status
      @headers = headers
    end

    # Parses the response body as JSON.
    # @return [Hash, Array]
    def json
      @json ||= begin
        Security.check_body_size!(@body, Security::MAX_RESPONSE_BODY_BYTES)
        JSON.parse(@body)
      end
    end

    # Returns whether the response was successful (2xx).
    # @return [Boolean]
    def success?
      status >= 200 && status < 300
    end
  end
end

File: go/pkg/fizzy/http.go
package fizzy

import (
	"context"
	"net/http"
	"time"
)

// Default values for HTTP client configuration.
// These can be overridden using functional options.
const (
	DefaultMaxRetries = 3
	DefaultBaseDelay  = 1 * time.Second
	DefaultMaxJitter  = 100 * time.Millisecond
	DefaultTimeout    = 30 * time.Second
	DefaultMaxPages   = 10000
)

// HTTPOptions configures the HTTP client behavior.
type HTTPOptions struct {
	// Timeout is the request timeout (default: 30s).
	Timeout time.Duration

	// MaxRetries is the maximum retry attempts for idempotent requests (default: 3).
	// POST requests only get 1 retry after successful token refresh.
	MaxRetries int

	// BaseDelay is the initial backoff delay (default: 1s).
	BaseDelay time.Duration

	// MaxJitter is the maximum random jitter to add to delays (default: 100ms).
	MaxJitter time.Duration

	// MaxPages is the maximum pages to fetch in GetAll (default: 10000).
	MaxPages int

	// Transport is the HTTP transport to use. If nil, a default transport
	// with sensible connection pooling is created.
	Transport http.RoundTripper
}

// DefaultHTTPOptions returns HTTPOptions with sensible defaults.
func DefaultHTTPOptions() HTTPOptions {
	return HTTPOptions{
		Timeout:    DefaultTimeout,
		MaxRetries: DefaultMaxRetries,
		BaseDelay:  DefaultBaseDelay,
		MaxJitter:  DefaultMaxJitter,
		MaxPages:   DefaultMaxPages,
	}
}

// WithTimeout sets the HTTP request timeout.
func WithTimeout(d time.Duration) ClientOption {
	return func(c *Client) {
		c.httpOpts.Timeout = d
	}
}

// WithMaxRetries sets the maximum number of retry attempts for idempotent requests.
func WithMaxRetries(n int) ClientOption {
	return func(c *Client) {
		c.httpOpts.MaxRetries = n
	}
}

// WithBaseDelay sets the initial backoff delay.
func WithBaseDelay(d time.Duration) ClientOption {
	return func(c *Client) {
		c.httpOpts.BaseDelay = d
	}
}

// WithMaxJitter sets the maximum random jitter to add to delays.
func WithMaxJitter(d time.Duration) ClientOption {
	return func(c *Client) {
		c.httpOpts.MaxJitter = d
	}
}

// WithMaxPages sets the maximum pages to fetch in GetAll.
func WithMaxPages(n int) ClientOption {
	return func(c *Client) {
		c.httpOpts.MaxPages = n
	}
}

// WithTransport sets a custom HTTP transport.
func WithTransport(t http.RoundTripper) ClientOption {
	return func(c *Client) {
		c.httpOpts.Transport = t
	}
}

// retryableError wraps an error with retry metadata.
// This allows respecting Retry-After headers from 429 responses.
type retryableError struct {
	err        error
	retryAfter time.Duration
}

func (r *retryableError) Error() string {
	return r.err.Error()
}

func (r *retryableError) Unwrap() error {
	return r.err
}

// newDefaultTransport creates an HTTP transport with sensible defaults.
// It clones http.DefaultTransport to preserve proxy settings, HTTP/2, TLS config.
func newDefaultTransport() http.RoundTripper {
	t := http.DefaultTransport.(*http.Transport).Clone()
	t.MaxIdleConns = 100
	t.MaxIdleConnsPerHost = 10
	t.IdleConnTimeout = 90 * time.Second
	return t
}

// noRetryKey is the context key for disabling retry on a request.
type noRetryKey struct{}

// WithNoRetry returns a context that disables retry for the request.
func WithNoRetry(ctx context.Context) context.Context {
	return context.WithValue(ctx, noRetryKey{}, true)
}

func isNoRetry(ctx context.Context) bool {
	v, _ := ctx.Value(noRetryKey{}).(bool)
	return v
}

// attemptKey is the context key for tracking request attempt number.
type attemptKey struct{}

// contextWithAttempt adds the request attempt number to the context.
func contextWithAttempt(ctx context.Context, attempt int) context.Context {
	return context.WithValue(ctx, attemptKey{}, attempt)
}

// attemptFromContext extracts the attempt number from context (defaults to 1).
func attemptFromContext(ctx context.Context) int {
	if v := ctx.Value(attemptKey{}); v != nil {
		if attempt, ok := v.(int); ok {
			return attempt
		}
	}
	return 1
}

// loggingTransport wraps an http.RoundTripper to log requests and responses,
// and calls observability hooks for all HTTP requests (including generated client).
type loggingTransport struct {
	inner  http.RoundTripper
	client *Client
}

// RoundTrip implements http.RoundTripper with logging and hooks.
func (t *loggingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	info := RequestInfo{
		Method:  req.Method,
		URL:     req.URL.String(),
		Attempt: attemptFromContext(req.Context()),
	}
	hookCtx := t.client.hooks.OnRequestStart(req.Context(), info)
	startTime := time.Now()

	req = req.WithContext(hookCtx)

	var result RequestResult
	defer func() {
		result.Duration = time.Since(startTime)
		t.client.hooks.OnRequestEnd(hookCtx, info, result)
	}()

	if t.client.logger != nil {
		t.client.logger.Debug("http request",
			"method", req.Method,
			"url", req.URL.String())
	}

	resp, err := t.inner.RoundTrip(req)

	if err != nil {
		result.Error = err
	} else {
		result.StatusCode = resp.StatusCode
		if resp.StatusCode == 429 || resp.StatusCode == 503 {
			result.RetryAfter = parseRetryAfter(resp.Header.Get("Retry-After"))
		}
		if t.client.logger != nil {
			t.client.logger.Debug("http response",
				"status", resp.StatusCode)
		}
	}

	return resp, err
}
